File: generator/src/main/groovy/ch/bailu/gtk/converter/NamespaceType.java
package ch.bailu.gtk.converter;

import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class NamespaceType {
    private final static Pattern NAMESPACE = Pattern.compile("^([A-Za-z]+)\\.([A-Za-z]+)$");
    private final static Pattern NONAMESPACE = Pattern.compile("^[A-Za-z]+$");


    private final String namespace;
    private final String type;

    private final boolean hasCurrentNamespace;
    private final boolean isValid;


    public NamespaceType(String currentNamespace, String type) {
        if (currentNamespace == null) {
            currentNamespace = "";
        }

        currentNamespace = currentNamespace.toLowerCase();

        if (type == null) {
            type = "";
        }



        Matcher m = NAMESPACE.matcher(type);

        if (m.find()) {
            namespace = m.group(1).toLowerCase();
            this.type = m.group(2);
            isValid = true;
        } else {
            m = NONAMESPACE.matcher(type);
            if (m.find()) {
                namespace = currentNamespace;
                this.type = type;
                isValid = true;
            } else {
                namespace = "";
                this.type = "";
                isValid = false;
            }
        }

        hasCurrentNamespace = namespace.equals(currentNamespace);
    }

    public boolean isValid() {
        return isValid;
    }

    public boolean hasCurrentNamespace() {
        return hasCurrentNamespace;
    }

    public String getNamespace() {
        return namespace;
    }

    public String getName() {
        return type;
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/model/type/NamespaceType.kt
package ch.bailu.gtk.model.type

import java.util.*
import java.util.regex.Pattern

class NamespaceType(fallbackNamespace: String, typeName: String) {
    companion object {
        private val NAMESPACE = Pattern.compile("^([A-Za-z_]\\w+)\\.([A-Za-z_]\\w+)$")
        private val NONAMESPACE = Pattern.compile("^[A-Za-z_]\\w+$")

        val INVALID = NamespaceType("", "")
    }

    val namespace: String
    val name: String

    val valid: Boolean
        get() = name.isNotEmpty()

    init {
        var m = NAMESPACE.matcher(typeName)
        if (m.find()) {
            namespace = m.group(1).lowercase()
            name = m.group(2)
        } else {
            m = NONAMESPACE.matcher(typeName)
            if (m.find()) {
                namespace = fallbackNamespace.lowercase()
                name = typeName
            } else {
                namespace = ""
                name = ""
            }
        }
    }

    override fun hashCode(): Int {
        return Objects.hash(namespace, name)
    }

    override fun toString(): String {
        return "${namespace}.${name}"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as NamespaceType

        if (namespace != other.namespace) return false
        if (name != other.name) return false

        return true
    }

    fun isCurrentNameSpace(currentNameSpace: String): Boolean {
        return namespace == currentNameSpace
    }
}
